### DISCUSSION TITLE
Nullability headaches with IComparer and IEqualityComparer

### DISCUSSION THREAD

**Discussion by jhudsoncedaron** [dotnet/runtime#103787] [Q&A]
We currently have:
```
public interface IComparer<T> {
    int Compare(T? x, T? y);
}

public interface IEqualityComparer<T> {
    bool Equals(T? x, T? y);
    int GetHashCode(T item);
}
```

The nullable annotations are messed up in the stock methods.

Demonstration:

```
   List<string> GetOrderedListOfItems(HashSet<string> items)
       => items.OrderBy(new ItemComparer()).ToList());

   class ItemComparer : IComparer<string>() {
       int IComparer<string>.Compare(string x, string y) { // Declaration has useless compiler warning. x and y are actually staticly typed to be not possibly null.
          // implement
       }
   }
```

IEqualityComparer has the opposite problem as well:

```
   // DictionaryFactory returns an IDictionary that behaves like Dictionary but with the null key check knocked out.
   IDictionary<string?, string?> container = DictonaryFactory<string?, string?>(new ItemEqualityComparer());
   container.Add(null, null); // Actually works if the IEqualityComparer handles null GetHashCode(!)

   class ItemComparer : IEqualityComparer<string?>() {
       int IEqualityComparer<string?>.GetHashCode(string? item) { // Useless compiler warning
          // implement
       }

       bool IEqualityComparer<string?>.Equals(string? x, string? y) {
          // implement
       }
   }
```

Both of these came up in real code multiple times. We've been slow to adapt nullability checks due to false alarms and false IDE automated fixes to them (but the latter is another story).

In general, these two are the same problem. I had both of these come up in live code with these exact use cases.

The problem is the nullability of the arguments doesn't correspond to the nullability of the interface. If the enumerable cannot contain a null; its OrderBy() comparer need not support nulls; and the inverse: if the dictionary key is allowed to be null then IEqualityComparer.GetHashCode() can receive a null. There's no way to shut up the compiler other than manually suppressing the warnings because _the warnings are wrong_ because nullability needs to actually follow the type.

**Comment by dotnet-policy-service** [comment]
Tagging subscribers to this area: @dotnet/area-system-collections
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.

**Comment by eiriktsarpalis** [comment]
> Declaration has useless compiler warning. x and y are actually staticly typed to be not possibly null.

Why do you think it's useless? It indicates that if you were to pass your implementation to a collection that does run equality comparisons against null values, your code wouldn't be able to handle them. The contract for IEqualityComparer<T> expects that null values should be accepted.

**Reply by jhudsoncedaron** [reply]
>    It indicates that if you were to pass your implementation to a collection that does run equality comparisons against null values

That would be a List<string?> not a List<string>; and therefore its comparer would be IComparer<string?> not IComparer<string>.

**Reply by stephentoub** [reply]
The annotations simply reflect the intended and documented behavior of the interface, which is that IEqualityComparer.Equals must allow for null inputs of reference types and that GetHashCode needn't permit them. You can disagree with the behavior that was defined for these many years ago, but both the docs and the annotations are correctly reflecting that.

**Reply by jhudsoncedaron** [reply]
In this case it's pretty obvious.

`.Sort()` and `.OrderBy()` had better not pass anything that's not in the collection and had better pass everything that's in the collection to the comparer. The only case of passing something not in the collection to the comparer is a binary tree like object such as `SortedDictionary` or `SortedSet`; and `new SortedSet<string>().Contains(null)` does indeed raise a nullability warning.

The actual behavior is pass the object through the collection type to the comparer type; if the documentation says otherwise, the documentation is wrong.

You see, I'm talking about useless compiler warnings; the actual rule is "the comparer gets passed exactly the objects that interact with the collection"; so when I write the code to that and get compiler warnings, the compiler warnings are wrong. This is the most recent case:

```
    private sealed class VersionComparer : IComparer<string>
    {
        // IEnumerable<string>.OrderBy() can't pass null without violating the nullability constraint of the collection so we know this isn't null.
#pragma warning disable CS8767
        public int Compare(string x, string y)
#pragma warning restore CS8767
        {
            int xbefore = 0, ybefore = 0;
            while (xbefore < x.Length || ybefore < y.Length)
            {
                int xafter = GetAfter(x, xbefore);
                int yafter = GetAfter(y, ybefore);
                int xv = GetComponentValue(x, xbefore, xafter);
                int yv = GetComponentValue(y, ybefore, yafter);
                if (xv < yv) return -1;
                if (xv > yv) return 1;
                xbefore = xafter;
                ybefore = yafter;
            }
            return 0;
        }
```

Like most custom logic comparers, this one is referenced exactly once with one set of well-formed input already filtered and most of the way through the processing steps; if anything were invalid it would have thrown long before this.

**Reply by jhudsoncedaron** [reply]
I tracked down StringComparer itself; despite what the documentation says, `StringComparer` does not implement `IComparable<string>` it implements `IComparable<string?>`

A comparer of type `string?` can be implicitly cast to a comparer of type `string`; however the inverse is not true; thus we are already in the world where nullability of comparer types has to be correct. Forcing the T? has no use.

**Reply by stephentoub** [reply]
> Forcing the T? has no use.

It does have use. It tells the implementer of the comparer implementation that it must support nulls, because that's the contract. It's been the contract long before C# added optional warnings around nullable reference types, and the annotations are enforcing that contract.

Again, you may disagree with the contract, and the documentation of that contract, but it's the contract none-the-less. You're choosing to violate that contract in your VersionComparer by null ref'ing on nulls, and that's your choice and a fine choice if you're just using this in your own code. But if you were to publish your VersionComparer for arbitrary consumption, there's a good chance it'd be used by folks that wouldn't get any warnings when they use it with collections that can contain null (e.g. because they don't have warnings for nullable reference types enabled), and your comparer not following the contract would result in their valid code blowing up.

**Reply by jhudsoncedaron** [reply]
If you are writing a comparer that can be used with nulls you always have to implement IComparer<T?> rather than IComparer<T> otherwise somebody who turns on nullable checks can't even use it.

By your own argument "It tells the implementer of the comparer implementation that it must support nulls" I expect zero cases of implementing `IComparer<T>` as opposed to `T?` in the standard libraries that are reference types.

**Reply by stephentoub** [reply]
> If you are writing a comparer that can be used with nulls

_Every_ comparer is supposed to be usable with null. That's the point.

**Reply by jhudsoncedaron** [reply]
Rather, Every _standard_ comparer is supposed to be usable with null. When you have a comparer for which it's ill defined to compare an element in most of the space the type can store, handling nulls is completely superfluous.

**Reply by chrisoverzero** [reply]
> You see, I'm talking about useless compiler warnings; the actual rule is "the comparer gets passed exactly the objects that interact with the collection" […]

> can't pass null without violating the nullability constraint of the collection so we know this isn't null.

What collection?

```csharp
StringComparer.OrdinalIgnoreCase.Equals(userInput, "activate") // "Activate" ==> true
```

**Reply by jhudsoncedaron** [reply]
@chrisoverzero 
```
   string? userInput = null;
   (new VersionComparer).Compare(userInput, "0.0.0.0")
```
yields the expected compiler warning. StringComparer implements `IComparable<string?>`. VersionComparer (my class) implements `IComparable<string>`.

**Comment by RenderMichael** [comment]
Defense in depart is always a good thing. There’s no guarantee a null didn’t make it in the `List<string>`, especially if it got de-serialized or called by a null-oblivious caller. Best to handle the null scenario.

**Comment by jhudsoncedaron** [comment]
@RenderMichael : My most immediate encounter of it was a case where any object not in the list being sorted would be doomed to failure as I actually have to parse the string to order by it; checking for null doesn't really help.

**Comment by RenderMichael** [comment]
You can turn a hard-to-diagnose `NullReferenceException` into a meaningful `InvalidOperationException` that has an explanatory message.

**Comment by jhudsoncedaron** [comment]
Hint to anybody trying to answer this: if you answer doesn't handle the IEqualityComparer.GetHashCode case where the static typing says it can't be null but it definitely can be null (I can put a breakpoint on the return statement in `if (item is null) return -1;` and it will be reached); it's not an answer.

So far everybody's trying to say IComparer.Compare should always handle nulls; think about what the equivalent to that is for  IEqualityComparer.GetHashCode ; and no changing the interface type is not an answer there; most of the time these are stock implementations of IEqualityComparer.

**Comment by RenderMichael** [comment]
It sounds like the disagreement here is with the nullability annotations placed on two specific types in the BCL, not with NRT broadly.

You can make a proposal to change the annotations for`IComparer.Compare` and `IEqualityComparer.Equals`. One option is to make them nullable-oblivious, like the return type for the non-generic `IEnumerable.GetEnumerator()`.

The proposal could look something like this:

```csharp
namespace System.Collections.Generic

public interface IComparer<T>
{
    public int Compare(
#nullable disable
        T x,
        T y
#nullable restore
    );
}

public interface IEqualityComparer<T>
{
    public bool Equals(
#nullable disable
        T x,
        T y
#nullable restore
    );
}
```

**Reply by jhudsoncedaron** [reply]
Ok now we're talking about something. I think that's a bit of a step backwards; I would totally expect intantiating IComparer<T> with `string` to have a `Compare(string, string)` but a `string?` to have a `Compare(string?, string?)` so the null-ness through the T that is being implemented.

Both of these types have explicit markings to change the rules away from that. `IComparer` literally says `int Compare(T? x, T? y);` ; almost as though it's for an older version of nullable where it didn't flow right through generic types.

**Reply by RenderMichael** [reply]
Feel free to make an API proposal.

**Comment by elgonzo** [comment]
> `if (item is null) return -1;`

> think about what the equivalent to that is for IEqualityComparer.GetHashCode

Well, according to the IEqualityComparer\<T\>.GetHashCode documentation, an implementation is supposed to throw an ArgumentNullException when a null reference is passed to it: https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.iequalitycomparer-1.gethashcode#exceptions

**Reply by jhudsoncedaron** [reply]
That document was wrong when it was first written in .NET Framework 2.0 and it's wrong today. Just because a stock implementation of IEqualityComparer can't handle a null doesn't mean a custom one can't.

I have a local sensible meaning for a null key, and it makes sense in context; and it makes sense to use it as the key for an IDictionary. Since I'm not going to reduplicate the IDictionary implementation again after writing it once, it needs to take stock IEqualityComparar, and will pass null to whatever IEqualityComparer it has it if it gets a null.

If I hadn't had to rewrite a new IDictionary implementation anyway I'd be requesting the check in Dictionary be knocked out and just let the one in EqualityComparer.Default handle it.
